command/data_format_test.go

```diff
@@ -13,56 +13,64 @@ type testData struct {
 	Name   string
 }
 
+var tData = testData{"global", "1", "example"}
+
 const expectJSON = `{
     "ID": "1",
     "Name": "example",
     "Region": "global"
 }`
 
-var (
-	tData        = testData{"global", "1", "example"}
-	testFormat   = map[string]string{"json": "", "template": "{{.Region}}"}
-	expectOutput = map[string]string{"json": expectJSON, "template": "global"}
-)
-
-func TestDataFormat(t *testing.T) {
-	ci.Parallel(t)
-	for k, v := range testFormat {
-		fm, err := DataFormat(k, v)
-		if err != nil {
-			t.Fatalf("err: %v", err)
-		}
-
-		result, err := fm.TransformData(tData)
-		if err != nil {
-			t.Fatalf("err: %v", err)
-		}
-
-		if result != expectOutput[k] {
-			t.Fatalf("expected output:\n%s\nactual:\n%s", expectOutput[k], result)
-		}
-	}
+var tcs = map[string]struct {
+	format   string
+	template string
+	expect   string
+	isError  bool
+}{
+	"json_good": {
+		format:   "json",
+		template: "",
+		expect:   expectJSON,
+	},
+	"template_good": {
+		format:   "template",
+		template: "{{.Region}}",
+		expect:   "global",
+	},
+	"template_bad": {
+		format:   "template",
+		template: "{{.foo}}",
+		isError:  true,
+		expect:   "can't evaluate field foo",
+	},
+	"template_empty": {
+		format:   "template",
+		template: "",
+		isError:  true,
+		expect:   "template needs to be specified in golang's text/template format.",
+	},
+	"template_sprig": {
+		format:   "template",
+		template: `{{$a := 1}}{{ $a | sprig_add 1 }}`,
+		expect:   "2",
+	},
 }
 
-func TestInvalidJSONTemplate(t *testing.T) {
+func TestDataFormat(t *testing.T) {
 	ci.Parallel(t)
-	// Invalid template {{.foo}}
-	fm, err := DataFormat("template", "{{.foo}}")
-	if err != nil {
-		t.Fatalf("err: %v", err)
-	}
-	_, err = fm.TransformData(tData)
-	if !strings.Contains(err.Error(), "can't evaluate field foo") {
-		t.Fatalf("expected invalid template error, got: %s", err.Error())
-	}
-
-	// No template is specified
-	fm, err = DataFormat("template", "")
-	if err != nil {
-		t.Fatalf("err: %v", err)
-	}
-	_, err = fm.TransformData(tData)
-	if !strings.Contains(err.Error(), "template needs to be specified the golang templates.") {
-		t.Fatalf("expected not specified template error, got: %s", err.Error())
+	for name, tc := range tcs {
+		t.Run(name, func(t *testing.T) {
+			ci.Parallel(t)
```

Comment:
```suggestion
```

I don't remember what happened, but I think this can cause some problems. Better to have the entire test function to be parallel rather than individual test cases.

Reply:
If you don't use `t.Parallel` in both the test and the subtest they don't parallelize like you'd think ([ref](https://deepsource.io/directory/analyzers/go/issues/GO-W6007)).  The issue I think you're recalling was around cases where we'd call `t.Parallel` in a test helper and then unknowingly call it again in the test that used that helper (without being inside a t.Run of its own). That will cause the panic about calling t.Parallel more than once.

Reply:
Ah interesting. TIL. Thanks!